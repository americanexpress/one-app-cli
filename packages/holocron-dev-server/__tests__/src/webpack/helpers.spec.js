/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import {
  createOneAppExternals,
  createHolocronModuleEntries,
} from '../../../src/webpack/helpers';

jest.mock('webpack', () => {
  const mockWebpack = jest.fn();
  mockWebpack.version = '';
  return mockWebpack;
});

describe('createOneAppExternals', () => {
  test('returns the set of externals used by One App', () => {
    const externals = createOneAppExternals();
    expect(externals).toMatchSnapshot();
  });

  test('returns the set of externals used by One App including added externals', () => {
    const externals = createOneAppExternals(['react-package']);
    expect(externals).toMatchSnapshot();
  });

  test('returns the set of externals from added externals in array formation [name, varName]', () => {
    const externals = createOneAppExternals([['react-package', 'reactPackage']]);
    expect(externals).toMatchSnapshot();
  });
});

describe('createHolocronModuleEntries', () => {
  test('returns an empty object if no modules were supplied', () => {
    const entries = createHolocronModuleEntries();
    expect(entries).toEqual({});
  });

  test('returns the entries (with hot entries) of modules to be bundled', () => {
    const moduleName = 'hot-module';
    const modulePath = '/some/path/to/hot-module';
    const modules = [{ moduleName, modulePath }];
    const entries = createHolocronModuleEntries({ modules, hot: true });
    expect(entries).toEqual({
      [moduleName]: [
        require.resolve('webpack-hot-middleware/client'),
        require.resolve('react-refresh/runtime'),
        `${modulePath}/src/index.js`,
      ],
    });
  });

  test('only returns the module entry when bundling without hot configured', () => {
    const moduleName = 'hot-module';
    const modulePath = '/some/path/to/hot-module';
    const modules = [{ moduleName, modulePath }];
    const entries = createHolocronModuleEntries({ modules });
    expect(entries).toEqual({
      [moduleName]: [`${modulePath}/src/index.js`],
    });
  });
});
