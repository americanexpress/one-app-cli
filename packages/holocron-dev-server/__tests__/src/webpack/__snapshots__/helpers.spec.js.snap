// Jest Snapshot v1, https://goo.gl/fbAQLP
// eslint-disable-next-line jest/no-large-snapshots
exports['createOneAppExternals returns the set of externals used by One App 1'] = `
Object {
  "@americanexpress/one-app-ducks": Object {
    "commonjs2": "@americanexpress/one-app-ducks",
    "root": "OneAppDucks",
    "var": "OneAppDucks",
  },
  "@americanexpress/one-app-router": Object {
    "commonjs2": "@americanexpress/one-app-router",
    "root": "OneAppRouter",
    "var": "OneAppRouter",
  },
  "create-shared-react-context": Object {
    "commonjs2": "create-shared-react-context",
    "root": "CreateSharedReactContext",
    "var": "CreateSharedReactContext",
  },
  "holocron": Object {
    "commonjs2": "holocron",
    "root": "Holocron",
    "var": "Holocron",
  },
  "holocron-module-route": Object {
    "commonjs2": "holocron-module-route",
    "root": "HolocronModuleRoute",
    "var": "HolocronModuleRoute",
  },
  "immutable": Object {
    "commonjs2": "immutable",
    "root": "Immutable",
    "var": "Immutable",
  },
  "prop-types": Object {
    "commonjs2": "prop-types",
    "root": "PropTypes",
    "var": "PropTypes",
  },
  "react": Object {
    "commonjs2": "react",
    "root": "React",
    "var": "React",
  },
  "react-dom": Object {
    "commonjs2": "react-dom",
    "root": "ReactDOM",
    "var": "ReactDOM",
  },
  "react-helmet": Object {
    "commonjs2": "react-helmet",
    "root": "ReactHelmet",
    "var": "ReactHelmet",
  },
  "react-redux": Object {
    "commonjs2": "react-redux",
    "root": "ReactRedux",
    "var": "ReactRedux",
  },
  "redux": Object {
    "commonjs2": "redux",
    "root": "Redux",
    "var": "Redux",
  },
  "reselect": Object {
    "commonjs2": "reselect",
    "root": "Reselect",
    "var": "Reselect",
  },
}
`;
